dashboard/src/modules/components/DatePickerComponent/index.jsx

```diff
@@ -28,12 +28,13 @@ function DatePickerWidget({
       : "To date must be greater than or equal to from date";
 
   const onFromChange = (_str, date) => {
-    setFromDate(date);
+    setFromDate(constructUTCDate(new Date(_str)));
     if (isValidDate(date)) {
       if (date > toDate) {
-        date.setDate(date.getDate() + 1);
-        setToDate(date);
-        setStrDate(formatDate(date));
+        let selectedDate = new Date(_str);
+        selectedDate.setDate(selectedDate.getDate() + 1);
+        setToDate(constructUTCDate(selectedDate));
+        setStrDate(formatDate(selectedDate));
```

Comment:
@webbnh That won't work.Let's take a look on how React setState works under the hood.
When a component renders,it encounters the useState() hook
`const [fromDate,setFromDate]=useState({})`

`useState `  returns 2 things,the value of the state for that particular render,and a setter function,to update the value of state for the **next** render of the component.

Now,whenever the setState() function is called,it does not update the value of state there and then (because it just can't ,and I will get back to why it can’t in just a moment).Think of setState() as a request to update the component with the new state value.It does not update it there and then,but instead remembers the value to which we want to update(by putting it inside a queue under the hood).

Now,when the component is rendered,React sees if the queue is empty or not,and if not,it re-renders the component with the updated values.

Now,why can't setState update the value of state there and then?Because the state for a particular render,is immutable.

Remember,when we first declared and initialized the state, we use **const**,and const values cannot be updated once initialized.But,when the component re-renders,a new instance of the const variable is created and is populated with the value from the queue.Now why react team decided to use const and not something like let is a subject I am not aware of (I just know some basic stuff).But yes,the above method proposed by you wont work since you are falling in the trap of reading the state immediately after setting the state.

Reply:
> whenever the setState() function is called, it does not update the value of state there and then

@shubham-html-css-js, I think I understand why that is (see below).

> Because the state for a particular render is immutable.

I'm afraid that I don't quite understand what you mean by that.  (That is, in my lexicon, "render" is what the user sees, but there also is what the Javascript code can access, and those might be different and I'm not sure which you mean; I'm also not clear on when re-renders happen.)


> we first declared and initialized the state, we use const, and const values cannot be updated once initialized.

The use of `const` means that the binding between the symbol and the value cannot be changed by the function.  This means that the "variable" cannot be reassigned.  In the case where the symbol is bound to a primitive type (like a string), then I you and I agree on the behavior.  E.g.,
```
  const myString = "yes";
  myString = "no";
```
is illegal.

However, in this case
```
  const myDate = new Date();
  myDate.setUTCDate(myDate.getUTCDate() + 1)
```
we are **not** _reassigning_ `myDate` -- `myDate` is bound to a reference to an object, and we're not changing that -- rather, we are modifying something else:  the object to which it refers.  So, this is actually legal code.

Thus, if the statement:
```
const [fromDate,setFromDate]=useState({})
```
creates a symbol `fromDate` which is bound to a reference to a `Date` object, and if calling `setFromDate()` modifies that object, then the modification should be immediately visible to `fromDate`.  However, if calling `setFromDate()` creates or supplies a _new_ `Date` object, then the update to `fromDate` won't be available until the next time the containing code is called and a new binding for `fromDate` is produced.

Does that all make sense?

(Note that the value passed to the update function (e.g., `setFromDate()`) [can be a callable](https://reactjs.org/docs/hooks-reference.html#functional-updates) which receives the current value of the state, which means that, if we wanted to, we probably _could_ cause the update to be made directly to the `Date` object which is already bound to `fromDate`, so the result would be visible immediately.)

Reply:
> That won't work.

So, let me amend my suggestion above.

```
    if (isValidDate(date)) {
      const selectedDate = constructUTCDate(new Date(_str));  // constructUTDDate() really should take _str, not a Date.
      setFromDate(selectedDate);
      if (selectedDate > toDate) {
        selectedDate.setUTCDate(date.getUTCDate() + 1);
        setToDate(selectedDate);
        setStrDate(formatDate(selectedDate));
```
That is, if Patternfly thinks the selected date is valid, then construct a UTC-date `Date` object from the _string_ which the user entered, and set that as the from-date; if the selected date is greater than the to-date, bump the selected date forward by one day (UTC), and use that result to set the to-date and the string representation of it.

This is basically what you were doing except:
- we only construct the `Date` object from the string once (and then we use it three times)
- we only construct the `Date` if the selected date is valid
- we only set `fromDate` if the selected date is valid
- we don't use Patternfly's `date` value for anything other than determining validity
- we use the UTC getter functions for manipulating the `Date` value.

Reply:
> So, let me amend my suggestion above.

This also fails.I tried it and eventually what this piece of code do is quite interesting.When we are changing selectedDate using `selectedDate.setUTCDate(date.getUTCDate() + 1)`,the object pointed by selectedDate changes.From Date was expecting the value from object pointed by selectedDate,but now that object has changed,so what is eventually happening is that when the component is re-rendered,both fromDate and toDate are getting the same values(which is one day greater then the selected fromDate),resulting in wrong output.

Reply:
As we discussed, my suggestion has a small flaw in that the `Date` object passed to `setFromDate` is passed as a reference, and therefore the subsequent change to it via `selectedDate` changes _both_ the `fromDate` and the `toDate`, since they are then both references to the same object.  😞

However, this is easily fixed by instantiating a new `Date` object for the `toDate` from the `Date` object used for the `fromDate`.  (The instantiation of a new `Date` from an existing `Date` should be fairly efficient, and it saves us from having to reparse the string, re-offsetting to midnight, and re-offsetting to UTC.)

Reply:
The third time is charmed?  😀 

```
    if (!isValidDate(date)) {
      setFromDate({});  // Since you think this makes the UI work better.
    } else {
      const selectedDate = constructUTCDate(_str);  // Assuming you've modified constructUTDDate()
      setFromDate(selectedDate);
      if (selectedDate > toDate) {
        const selectedToDate = new Date(selectedDate);
        selectedToDate.setUTCDate(selectedToDate.getUTCDate() + 1);
        setToDate(selectedToDate);
        setStrDate(formatDate(selectedToDate));
```
That is, we use a single `if` statement to deal with whether the from-date is valid (setting `fromDate` to a generic object, if it's not); we let `constructUTCDate()` instantiate the `Date` object, instead of doing that locally and having it modify or reinstantiate it; _if_ we need to update the to-date, we copy the from-date to a new `Date` object, modify that, and set it.

As we discussed, we may need to do some management of the displayed to-date value, so in this code we might want to set `strDate` to `selectedDate` instead of `selectedToDate`.  And, you might want to refactor the code which bumps the to-date forward by a day into a helper function which you can also use when the user selects the to-date.